hooks/useBuyNft.ts

```diff
@@ -0,0 +1,153 @@
+import { useCallback, useEffect, useState } from "react";
+import {
+  useConnect,
+  useWaitForTransactionReceipt,
+  useWriteContract,
+  useAccount
+} from "wagmi";
+import { useToast } from "@chakra-ui/react";
+import psycSaleAbiSepolia from "../abis/psycSaleAbiSepolia.json";
+import { psycSaleSepolia } from "../constants/contracts";
+// import { parseUnits } from "viem";
+
+type ArgsType =
+  | [number, string[]]
+  | [number]
+  | [number, number, string[]]
+  | [number, number];
+
+const useBuyNft = (isPrivateSale: boolean, isRandom: boolean) => {
+  const { isConnected } = useAccount();
+  const { connect, connectors } = useConnect();
+  const toast = useToast();
+  const [isMinting, setIsMinting] = useState(false);
+  const [isConfirmed, setIsConfirmed] = useState(false);
+  const [isModalOpen, setIsModalOpen] = useState(false);
+
+  const { data: hash, writeContract, isPending, error } = useWriteContract();
+
+  const { isLoading: isConfirming, isSuccess } = useWaitForTransactionReceipt({
+    hash
+  });
+
+  const buyNft = useCallback(
+    async (
+      batchId: number,
+      erc721TokenId: number,
+      price: string,
+      proof: string[] = []
+    ) => {
+      if (!isConnected) {
+        toast({
+          title: "Please connect your wallet first",
+          position: "top-right",
+          status: "error",
+          isClosable: true
+        });
+        if (connectors[0]) {
+          connect({ connector: connectors[0] });
+        }
+        return;
+      }
+
+      try {
+        setIsMinting(true);
+        let functionName = "";
+        let args: ArgsType = [batchId, erc721TokenId];
+
+        if (isPrivateSale && isRandom) {
+          functionName = "buyRandomFromBatch";
+          args = [batchId, proof];
+        } else if (!isPrivateSale && isRandom) {
+          functionName = "buyRandomNftCopyFromBatch";
+          args = [batchId];
+        } else if (isPrivateSale && !isRandom) {
+          functionName = "buyFromBatch";
+          args = [batchId, erc721TokenId, proof];
+        } else if (!isPrivateSale && !isRandom) {
+          functionName = "buyNftCopyFromBatch";
+          args = [batchId, erc721TokenId];
+        }
+        // const priceInWei = parseUnits(price?.split(" ")[0] ?? "0", 18);
+
+        writeContract({
+          address: psycSaleSepolia,
+          abi: psycSaleAbiSepolia,
+          functionName: functionName,
+          args: args
+          // value: priceInWei
```

Comment:
in case we dont need the value(we might need it) pls remove this.

Reply:
alright